uqcsbot/scripts/meme.py

```diff
@@ -0,0 +1,192 @@
+    "sarcasticbear": "Sarcastic Bear",
+    "sb": "Scumbag Brain",
+    "scc": "Sudden Clarity Clarence",
+    "sf": "Sealed Fate",
+    "sk": "Skeptical Third World Kid",
+    "ski": "Super Cool Ski Instructor",
+    "snek": "Skeptical Snake",
+    "soa": "Seal of Approval",
+    "sohappy": "I Would Be So Happy",
+    "sohot": "So Hot Right Now",
+    "sparta": "This is Sparta!",
+    "spongebob": "Mocking Spongebob",
+    "ss": "Scumbag Steve",
+    "stew": "Baby, You've Got a Stew Going",
+    "success": "Success Kid",
+    "tenguy": "10 Guy",
+    "toohigh": "The Rent Is Too Damn High",
+    "tried": "At Least You Tried",
+    "ugandanknuck": "Ugandan Knuckles",
+    "whatyear": "What Year Is It?",
+    "winter": "Winter is coming",
+    "wonka": "Condescending Wonka",
+    "xy": "X all the Y",
+    "yallgot": "Y'all Got Any More of Them",
+    "yodawg": "Xzibit Yo Dawg",
+    "yuno": "Y U NO Guy",
+}
+
+# TODO: Would be really simple to add custom UQCS memes
+
+
+@bot.on_command("meme")
+@loading_status
+def handle_meme(command: Command):
+    """
+    !meme <meme name> "<top text>" "<bottom text>"
+
+    For a full list of names: !meme names
+    """
+    channel = command.channel_id
+
+    if not command.has_arg():
+        bot.post_message(channel, "Please run !help meme for usage")
+        return
+
+    name = command.arg.split()[0].lower()
+    if name != "names" and name not in MEME_NAMES.keys():
+        bot.post_message(channel, "The meme name is invalid. Try !meme names to get a list of all valid names")
+        return
+
+    args = get_meme_arguments(command.arg)
+
+    if name == "names":
+        send_meme_names(command)
+        return
+    elif len(args) == 2:
+        top, bottom = args
+    else:
+        bot.post_message(channel, "You supplied the wrong number of args. Please run !help meme")
+        return
+
+    # Make an attachment linking to image
+    image_url = API_URL + f"{quote(name)}/{quote(top)}/{quote(bottom)}.jpg"
+    attachments = [{"text": "", "image_url": image_url}]
+    bot.post_message(channel, "", attachments=attachments)
+
+
+@success_status
+def send_meme_names(command: Command):
+    """Sends the full list of meme names to the users channel to avoid channel spam"""
+    user_channel = bot.channels.get(command.user_id)
+    names_text = "\n".join((f"{full_name}: {name}" for (name, full_name) in MEME_NAMES.items()))
+    attachments = [{'text': names_text, 'title': "Meme Names:"}]
+    bot.post_message(user_channel, "", attachments=attachments)
+
+
+def get_meme_arguments(input_args: str):
+    """Gets the top and bottom text and returns them in a url friendly form that conforms with the api standards"""
+    # This gets the text between the quotation marks (and ignores \")
+    args = re.findall(r'"(.*?(?<!\\))"', input_args)
+
+    # Replaces all the required characters to be url friendly
+    url_friendly_args = []
+    for arg in args:
+        for old, new in REPLACEMENTS.items():
```

Comment:
https://www.tutorialspoint.com/python/string_translate.htm might be useful here :)

Reply:
Oh yeah, had that at one point but had to remove it for multiple character keys which are no longer needed. Will add back, cheers.